pyomo/repn/plugins/gams_writer.py

```diff
@@ -104,15 +119,122 @@ def visiting_potential_leaf(self, node):
         return False, None
 
 
-def expression_to_string(expr, labeler=None, smap=None):
+def expression_to_string(expr, model, labeler=None, smap=None):
     if labeler is not None:
         if smap is None:
             smap = SymbolMap()
         smap.default_labeler = labeler
-    visitor = ToGamsVisitor(smap)
+    visitor = ToGamsVisitor(smap, model)
     return visitor.dfs_postorder_stack(expr)
 
 
+class Categorizer(object):
+    """Class for representing categorized variables.
+
+    Given a list of variable names and a symbol map, categorizes the variable
+    names into the categories: binary, ints, positive and reals.
+
+    """
+
+    def __init__(self, var_list, symbol_map):
+        self.binary = []
+        self.ints = []
+        self.positive = []
+        self.reals = []
+
+        # categorize variables
+        for var in var_list:
+            v = symbol_map.getObject(var)
+            if v.is_binary():
+                self.binary.append(var)
+            elif v.is_integer():
+                if (v.has_lb() and (value(v.lb) >= 0)) and \
+                   (v.has_ub() and (value(v.ub) <= 1)):
+                    self.binary.append(var)
+                else:
+                    self.ints.append(var)
+            elif value(v.lb) == 0:
+                self.positive.append(var)
+            else:
+                self.reals.append(var)
+
+    def __iter__(self):
+        """Iterate over all variables.
+
+        Yield a tuple containing the variables category and its name.
+        """
+        for category in ['binary', 'ints', 'positive', 'reals']:
+            var_list = getattr(self, category)
+            for var_name in var_list:
+                yield category, var_name
+
+
+def split_long_line(line):
+    """
+    GAMS has an 80,000 character limit for lines, so split as many
+    times as needed so as to not have illegal lines.
+    """
+    new_lines = ''
+    while len(line) > 80000:
+        i = 80000
+        while line[i] != ' ':
+            # Walk backwards to find closest space,
+            # where it is safe to split to a new line
+            if i < 0:
+                raise RuntimeError(
+                    "Found an 80,000+ character string with no spaces")
+            i -= 1
+        new_lines += line[:i] + '\n'
+        line = line[i + 1:]
+    new_lines += line
+    return new_lines
+
+
+def check_on_model(model, comp):
```

Comment:
Why not make this a class?  Then it can store a ComponentSet of known blocks... it can initialize the ComponentSet with the model and all the model's parents.  Then `check_on_model` only has to check the object's parent_block().  If it is not in the set, walk up the hierarchy an collect "new" blocks.  If you encounter one that is in the ComponentSet, then add everything you've run across to the Set and return True; else if you get to the root without finding anything in the Set, throw the list of collected blocks away and return False.

Reply:
Good idea. Blocks are hashable though, so is there any reason not to just use a normal set?